File: export-scripts/EXP_metadata_json.R
###########################################################################################
###########################################################################################
##
## Building metadata.json and comparisons.json using nested DataFrames
##
###########################################################################################
###########################################################################################

#=========================================================================================#
# Setting up
#=========================================================================================#

#-----------------------------------------------------------------------------------------#
# Loading libraries
#-----------------------------------------------------------------------------------------#

suppressWarnings(suppressMessages(library(tidyverse)))
suppressWarnings(suppressMessages(library(DBI)))
suppressWarnings(suppressMessages(library(dbplyr)))
suppressWarnings(suppressMessages(library(odbc)))
suppressWarnings(suppressMessages(library(lubridate)))
suppressWarnings(suppressMessages(library(fs)))
suppressWarnings(suppressMessages(library(jsonlite))) # needs to be version 1.8.4
suppressWarnings(suppressMessages(library(rlang)))
suppressWarnings(suppressMessages(library(svDialogs)))
suppressWarnings(suppressMessages(library(scales)))

#-----------------------------------------------------------------------------------------#
# set summarise options
#-----------------------------------------------------------------------------------------#

options(dplyr.summarise.inform = FALSE)

#-----------------------------------------------------------------------------------------#
# get and set env vars
#-----------------------------------------------------------------------------------------#

# find script

set_environment_loc <-
    list.files(
        getwd(),
        pattern = "set_environment.R",
        full.names = TRUE,
        recursive = TRUE
    )

# run script

source(set_environment_loc)


#-----------------------------------------------------------------------------------------#
# Connect to database
#-----------------------------------------------------------------------------------------#

# determining driver to use (so script works across machines)

odbc_driver <- 
    odbcListDrivers() %>% 
    pull(name) %>% 
    unique() %>% 
    str_subset("ODBC Driver") %>% 
    sort(decreasing = TRUE) %>% 
    head(1)

# if no "ODBC Driver", use Windows built-in driver

if (length(odbc_driver) == 0) odbc_driver <- "SQL Server"

# using Windows auth with no DSN

EHDP_odbc <-
    dbConnect(
        drv = odbc::odbc(),
        driver = paste0("{", odbc_driver, "}"),
        server = server,
        database = db_name,
        trusted_connection = "yes",
        encoding = "utf8",
        trustservercertificate = "yes"
    )


#-----------------------------------------------------------------------------------------#
# create folders if they don't exist
#-----------------------------------------------------------------------------------------#

dir_create(path(base_dir, "indicators/metadata"))


#=========================================================================================#
# Pulling & writing data ----
#=========================================================================================#

#-----------------------------------------------------------------------------------------#
# all metadata
#-----------------------------------------------------------------------------------------#

EXP_metadata <- 
    EHDP_odbc %>% 
    tbl("EXP_metadata") %>% 
    collect() %>% 
    arrange(
        IndicatorID,
        MeasureID,
        TimePeriodID
    ) %>% 
    mutate(
        across(
            where(is.character),
            ~ as_utf8_character(enc2native(.x))
        )
    )

#-----------------------------------------------------------------------------------------#
# general datasets for joining
#-----------------------------------------------------------------------------------------#

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# distinct measures
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

distinct_measures <-
    EXP_metadata %>%
    select(
        IndicatorID,
        MeasureID
    ) %>%
    distinct()


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# indicator and measure text fields
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

indicator_measure_text <-
    EXP_metadata %>%
    select(
        IndicatorID,
        IndicatorName,
        IndicatorLabel,
        IndicatorDescription,
        MeasureID,
        MeasureName,
        MeasurementType,
        how_calculated,
        Sources,
        DisplayType
    ) %>%
    distinct()


#-----------------------------------------------------------------------------------------#
# nesting vis options
#-----------------------------------------------------------------------------------------#

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# table options
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# TimePeriodID

measure_table_time <- 
    EXP_metadata %>% 
    filter(Table == 1) %>%
    select(
        IndicatorID,
        MeasureID,
        GeoType,
        TimePeriodID
    ) %>% 
    distinct() %>% 
    left_join(
        distinct_measures,
        .,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    group_by(IndicatorID, MeasureID, GeoType) %>% 
    summarise(TimePeriodID = list(unname(unlist(TimePeriodID)))) %>% 
    ungroup()


# combining

measure_table <- 
    left_join(
        distinct_measures,
        measure_table_time,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    group_by(IndicatorID, MeasureID) %>% 
    group_nest(.key = "Table", keep = FALSE)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# map options
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# RankReverse: 0/1

measure_mapping_rr <- 
    EXP_metadata %>% 
    filter(Map == 1) %>% 
    select(
        IndicatorID,
        MeasureID,
        RankReverse
    ) %>% 
    left_join(
        distinct_measures,
        .,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    group_by(MeasureID) %>% 
    summarise(RankReverse = max(RankReverse))

# TimePeriodID

measure_mapping_time <- 
    EXP_metadata %>% 
    filter(Map == 1) %>% 
    select(
        IndicatorID,
        MeasureID,
        GeoType,
        TimePeriodID
    ) %>% 
    distinct() %>% 
    left_join(
        distinct_measures,
        .,
        by = c("IndicatorID", "MeasureID")
    ) %>%     
    group_by(IndicatorID, MeasureID, GeoType) %>% 
    summarise(TimePeriodID = list(unname(unlist(TimePeriodID)))) %>% 
    ungroup()


# combining

measure_mapping <- 
    left_join(
        distinct_measures,
        measure_mapping_time,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    left_join(
        measure_mapping_rr,
        by = "MeasureID"
    ) %>% 
    group_by(IndicatorID, MeasureID) %>% 
    group_nest(.key = "Map", keep = FALSE)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# trend options
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# TimePeriodID

measure_trend_time <- 
    EXP_metadata %>% 
    filter(Trend == 1) %>% 
    select(
        IndicatorID,
        MeasureID,
        GeoType,
        TimePeriodID
    ) %>% 
    distinct() %>% 
    left_join(
        distinct_measures,
        .,
        by = c("IndicatorID", "MeasureID")
    ) %>%     
    group_by(IndicatorID, MeasureID, GeoType) %>% 
    summarise(TimePeriodID = list(unname(unlist(TimePeriodID)))) %>% 
    ungroup()


# combining

measure_trend <- 
    left_join(
        distinct_measures,
        measure_trend_time,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    group_by(IndicatorID, MeasureID) %>% 
    group_nest(.key = "Trend", keep = FALSE)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# measure (non-boro) trend comparisons
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# ==== specific comparisons view ==== #

EXP_comparisons <- 
    EHDP_odbc %>% 
    tbl("EXP_comparisons") %>% 
    collect() %>% 
    arrange(
        IndicatorID,
        ComparisonID,
        MeasureID
    )

# ==== nesting ComparisonIDs ==== #

indicator_comparisons <- 
    EXP_comparisons %>% 
    select(
        IndicatorID,
        Comparisons = ComparisonID
    ) %>% 
    distinct() %>% 
    drop_na() %>% 
    group_by(IndicatorID) %>% 
    summarise(Comparisons = list(unname(unlist(Comparisons))))


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# links and disparities
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# ==== specific link view ==== #

# because left-joining these 400 rows added 12k rows to the view

MeasureID_links <- 
    EHDP_odbc %>% 
    tbl("EXP_links") %>% 
    collect() %>% 
    arrange(
        BaseMeasureID,
        MeasureID
    )


# ==== nesting links ==== #

measure_links <- 
    MeasureID_links %>% 
    filter(disparity_flag == 0) %>% 
    select(-disparity_flag) %>% 
    distinct() %>% 
    left_join(
        distinct_measures %>% select(BaseMeasureID = MeasureID),
        .,
        by = "BaseMeasureID"
    ) %>% 
    group_by(BaseMeasureID) %>% 
    group_nest(.key = "Measures", keep = FALSE) %>% 
    rename(MeasureID = BaseMeasureID)


# ==== nesting disparities ==== #

# Disparities: 0/1

measure_disp <- 
    MeasureID_links %>% 
    filter(disparity_flag == 1) %>% 
    group_by(BaseMeasureID) %>% 
    summarise(Disparities = max(disparity_flag), .groups = "keep") %>% 
    rename(MeasureID = BaseMeasureID) %>% 
    left_join(
        distinct_measures %>% select(MeasureID),
        .,
        by = "MeasureID"
    ) %>% 
    mutate(Disparities = replace_na(Disparities, 0L))


# ==== nesting disparities ==== #

measure_links_disp <- 
    left_join(
        measure_links,
        measure_disp,
        by = "MeasureID"
    ) %>% 
    group_by(MeasureID) %>% 
    group_nest(.key = "Links", keep = FALSE)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# combining map, trend, and links, then nesting those under VisOptions
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

measure_vis_options <- 
    left_join(
        measure_table,
        measure_mapping,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    left_join(
        measure_trend,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    left_join(
        measure_links_disp,
        by = "MeasureID"
    ) %>% 
    group_by(IndicatorID, MeasureID) %>% 
    group_nest(.key = "VisOptions", keep = FALSE)


#-----------------------------------------------------------------------------------------#
# nesting geotypes
#-----------------------------------------------------------------------------------------#

measure_geotypes <- 
    EXP_metadata %>% 
    select(
        IndicatorID,
        MeasureID,
        AvailableGeoTypes = GeoType
    ) %>% 
    distinct() %>% 
    group_by(IndicatorID, MeasureID) %>% 
    summarise(AvailableGeoTypes = list(unname(unlist(AvailableGeoTypes)))) %>% 
    ungroup()


#-----------------------------------------------------------------------------------------#
# nesting times
#-----------------------------------------------------------------------------------------#

measure_times <- 
    EXP_metadata %>% 
    select(
        IndicatorID,
        MeasureID,
        AvailableTimePeriodIDs = TimePeriodID
    ) %>% 
    distinct() %>% 
    group_by(IndicatorID, MeasureID) %>% 
    summarise(AvailableTimePeriodIDs = list(unname(unlist(AvailableTimePeriodIDs)))) %>% 
    ungroup()


#-----------------------------------------------------------------------------------------#
# nesting times
#-----------------------------------------------------------------------------------------#

trend_no_compare <- 
    EXP_metadata %>% 
    select(
        IndicatorID,
        MeasureID,
        Sources
    ) %>% 
    distinct() %>% 
    transmute(
        IndicatorID,
        MeasureID,
        TrendNoCompare = 
            case_when(
                # CHS
                Sources %>% str_detect("New York City Community Health Survey") ~ "2021",
                # NYC Kids
                Sources %>% str_detect("NYC KIDS Survey") ~ "2021",
                .default = NA_character_

            )
    )


#-----------------------------------------------------------------------------------------#
# combining geotype, times, and vis options, then nesting those under other measure-level info vars
#-----------------------------------------------------------------------------------------#

metadata <- 
    left_join(
        indicator_measure_text,
        measure_geotypes,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    left_join(
        trend_no_compare,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    left_join(
        measure_times,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    left_join(
        measure_vis_options,
        by = c("IndicatorID", "MeasureID")
    ) %>% 
    group_by(
        IndicatorID,
        IndicatorName,
        IndicatorLabel,
        IndicatorDescription
    ) %>% 
    group_nest(.key = "Measures", keep = FALSE) %>% 
    left_join(
        indicator_comparisons,
        by = "IndicatorID"
    ) %>% 
    relocate(
        IndicatorID,
        IndicatorName,
        IndicatorLabel,
        IndicatorDescription,
        Comparisons,
        Measures
    )


#-----------------------------------------------------------------------------------------#
# saving files ----
#-----------------------------------------------------------------------------------------#

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# converting to JSON
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# also making null values empty

metadata_json_pretty <- 
    metadata %>% 
    toJSON(pretty = TRUE, null = "null", na = "null") %>% 
    str_replace_all("\\[null\\]", "[]")

metadata_json <- 
    metadata %>% 
    toJSON(pretty = FALSE, null = "null", na = "null") %>% 
    str_replace_all("\\[null\\]", "[]")

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# saving JSON
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

write_file(metadata_json_pretty, path(base_dir, "indicators/metadata/metadata_pretty.json"))
write_file(metadata_json,        path(base_dir, "indicators/metadata/metadata.json"))


#=========================================================================================#
# closing database connection ----
#=========================================================================================#

dbDisconnect(EHDP_odbc)

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #
# #                             ---- THIS IS THE END! ----
# #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

File: export-scripts/EXP_metadata_json.py
###########################################################################################
###########################################################################################
##
## Building metadata.json and comparisons.json using nested DataFrames
##
###########################################################################################
###########################################################################################

#=========================================================================================#
# Setting up
#=========================================================================================#

#-----------------------------------------------------------------------------------------#
# Loading libraries
#-----------------------------------------------------------------------------------------#

import pyodbc
import pandas as pd
import easygui
import os
import warnings
import re
import numpy as np

warnings.simplefilter("ignore")

#-----------------------------------------------------------------------------------------#
# get and set env vars
#-----------------------------------------------------------------------------------------#

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# base_dir for absolute path
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# get environemnt var

base_dir = os.environ.get("base_dir", "")

if (base_dir == ""):
    
    # get current folder
    
    this_dir = os.path.basename(os.path.abspath("."))
    
    # if the current folder is "EHDP-data", use the absolute path to it
    
    if (this_dir == "EHDP-data"):
        
        base_dir = os.path.abspath(".")
        
    else:
        
        # if the current folder is below "EHDP-data", switch it
        
        base_dir = re.sub(r"(.*EHDP-data)(.*)", r"\1", os.path.abspath("."))
        
    os.environ["base_dir"] = base_dir


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# server
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# get envionment var

server = os.environ.get("server", "")

if (server == ""):
        
    computername = os.environ.get("COMPUTERNAME", "")

    if (computername != "DESKTOP-PU7DGC1"):
        
        # ask and set
        
        server = "SQLIT04A"
        
        os.environ["server"] = server

    else:
            
        server = "DESKTOP-PU7DGC1"
        
        os.environ["server"] = server


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# database
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# get envionment var

data_env = os.environ.get("data_env", "")

if (data_env == ""):
    
    # ask and set
    
    data_env = easygui.enterbox("staging [s] or production [p]?")
    
    os.environ["data_env"] = data_env

# set DB name

if (data_env.lower() == "s"):
    
    # staging
    
    db_name = "BESP_IndicatorAnalysis"

elif (data_env.lower() == "p"):
    
    # production
    
    db_name = "BESP_Indicator"


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# determining which driver to use
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

drivers_list = pyodbc.drivers()

# odbc

odbc_driver_list = list(filter(lambda dl: "ODBC Driver" in dl, drivers_list))
odbc_driver_list.sort(reverse = True)

# native

native_driver_list = list(filter(lambda dl: "SQL Server Native Client" in dl, drivers_list))
native_driver_list.sort()

# deciding & setting

if len(odbc_driver_list) > 0:
    
    driver = odbc_driver_list[0]
    
elif len(native_driver_list) > 0:
    
    driver = native_driver_list[0]
    
else:
    
    driver = "SQL Server"

#-----------------------------------------------------------------------------------------#
# Connect to database
#-----------------------------------------------------------------------------------------#

EHDP_odbc = pyodbc.connect("DRIVER={" + driver + "};SERVER=" + server + ";DATABASE=" + db_name + ";Trusted_Connection=yes;trustservercertificate=yes")

#-----------------------------------------------------------------------------------------#
# create folders if they don't exist
#-----------------------------------------------------------------------------------------#

os.makedirs(base_dir + "/indicators/metadata/", exist_ok = True)


#=========================================================================================#
# data ops
#=========================================================================================#

EXP_metadata = (
    pd.read_sql("SELECT * FROM EXP_metadata", EHDP_odbc)
    .sort_values(by = ["IndicatorID", "MeasureID", "end_period"])
)

#-----------------------------------------------------------------------------------------#
# nesting vis options
#-----------------------------------------------------------------------------------------#

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# map options
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# map flag

measure_mapping_flag = (
    EXP_metadata
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "Map"
        ]
    ]    
    .drop_duplicates()
    .rename(columns = {"Map": "Map_flag"})
)

# RankReverse: 0/1

measure_mapping_rr = (
    EXP_metadata
    .query("Map == '1'")
    .loc[:,
        [
            "IndicatorID",
            "MeasureID",
            "RankReverse"
        ]
    ]
    .drop_duplicates()
    .groupby(["IndicatorID", "MeasureID"], dropna = False)
    .apply(lambda x: x[["RankReverse"]].drop_duplicates().to_dict("records")[0])
    .reset_index()
    .rename(columns = {0: "Map_rr"})
)

# TimeDescription

measure_mapping_time = (
    EXP_metadata
    .query("Map == '1'")
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "TimeDescription"
        ]
    ]    
    .drop_duplicates()
    .groupby(["IndicatorID", "MeasureID"], dropna = False)
    .apply(lambda x: x[["TimeDescription"]].drop_duplicates().to_dict("list"))
    .reset_index()
    .rename(columns = {0: "Map_time"})
)

# GeoType

measure_mapping_geo = (
    EXP_metadata
    .query("Map == '1'")
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "GeoType"
        ]
    ]    
    .drop_duplicates()
    .groupby(["IndicatorID", "MeasureID"], dropna = False)
    .apply(lambda x: x[["GeoType"]].drop_duplicates().to_dict("list"))
    .reset_index()
    .rename(columns = {0: "Map_geo"})
)

# combining

measure_mapping = (
    pd.merge(
        measure_mapping_flag,
        measure_mapping_time,
        how = "left"
    )
    .merge(
        measure_mapping_geo,
        how = "left"
    )
    .merge(
        measure_mapping_rr,
        how = "left"
    )
    # .assign(Map = lambda x: pd.DataFrame([x["Map_time"], x["Map_geo"], x["Map_rr"]]).to_dict("list"))
    # .assign(Map = lambda x: np.where(x.Map_flag == 0, None, x.Map))
    # .loc[:, 
    #     [
    #         "IndicatorID",
    #         "MeasureID",
    #         "Map"
    #     ]
    # ]
)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# trend options
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# measure_trend = (
#     EXP_metadata
#     .loc[:, 
#         [
#             "IndicatorID",
#             "MeasureID",
#             "Trend",
#             "Disparities"
#         ]
#     ]
#     .drop_duplicates()
#     .rename(columns = {"Trend": "On"})
#     .groupby(["IndicatorID", "MeasureID"], dropna = False)
#     .apply(lambda x: x[["On", "Disparities"]].to_dict("records"))
#     .reset_index()
#     .rename(columns = {0: "Trend"})
# )

# trend flag

measure_trend_flag = (
    EXP_metadata
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "Trend"
        ]
    ]    
    .drop_duplicates()
    .rename(columns = {"Trend": "Trend_flag"})
)


# TimeDescription

measure_trend_time = (
    EXP_metadata
    .query("Trend == '1'")
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "TimeDescription"
        ]
    ]    
    .drop_duplicates()
    .groupby(["IndicatorID", "MeasureID"], dropna = False)
    .apply(lambda x: x[["TimeDescription"]].drop_duplicates().to_dict("list"))
    .reset_index()
    .rename(columns = {0: "Trend_time"})
)

# GeoType

measure_trend_geo = (
    EXP_metadata
    .query("Trend == '1'")
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "GeoType"
        ]
    ]    
    .drop_duplicates()
    .groupby(["IndicatorID", "MeasureID"], dropna = False)
    .apply(lambda x: x[["GeoType"]].drop_duplicates().to_dict("list"))
    .reset_index()
    .rename(columns = {0: "Trend_geo"})
)

# combining

measure_trend = (
    pd.merge(
        measure_trend_flag,
        measure_trend_time,
        how = "left"
    )
    .merge(
        measure_trend_geo,
        how = "left"
    )
    .merge(
        measure_trend_disp,
        how = "left"
    )
    .assign(Trend = lambda x: pd.DataFrame([x["Trend_time"], x["Trend_geo"], x["Trend_disp"]]).to_dict("list"))
    .assign(Trend = lambda x: np.where(x.Trend_flag == 0, None, x.Trend))
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "Trend"
        ]
    ]  
)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# measure (non-boro) trend comparisons
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# ==== specific comparisons view ==== #

EXP_comparisons = (
    pd.read_sql("SELECT * FROM EXP_comparisons", EHDP_odbc)
    .sort_values(by = ["IndicatorID", "ComparisonID", "MeasureID"])
)


# ==== nesting ComparisonIDs ==== #

indicator_comparisons = (
    EXP_comparisons
    .loc[:, ["IndicatorID", "ComparisonID"]]
    .drop_duplicates()
    .dropna()
    .groupby(["IndicatorID"])
    .apply(lambda x: x["ComparisonID"].tolist()) # [] returns Series, not DataFrame
    .reset_index()
    .rename(columns = {0: "Comparisons"})
)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# linked measures
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# ==== specific link view ==== #

# because left-joining these 400 rows added 12k rows to the view

MeasureID_links = (
    pd.read_sql("SELECT * FROM EXP_links", EHDP_odbc)
    .loc[:, ["BaseMeasureID", "MeasureID", "SecondaryAxis"]]
    .sort_values(by = ["BaseMeasureID", "MeasureID"])
)

# ==== nesting links ==== #

measure_links = (
    MeasureID_links
    .drop_duplicates()
    .groupby(["BaseMeasureID"], dropna = False)
    .apply(lambda x: x[["MeasureID", "SecondaryAxis"]].to_dict("records"))
    .reset_index()
    .rename(columns = {0: "Links", "BaseMeasureID": "MeasureID"})
)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# disparities
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

# Disparities: 0/1

measure_disp = (
    EXP_metadata
    .query("Trend == '1'")
    .loc[:, 
        [
            "IndicatorID",
            "MeasureID",
            "Disparities"
        ]
    ]    
    .drop_duplicates()
    .groupby(["IndicatorID", "MeasureID"], dropna = False)
    .apply(lambda x: x[["Disparities"]].drop_duplicates().to_dict("records")[0])
    .reset_index()
    .rename(columns = {0: "Trend_disp"})
)


# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
# combining map, trend, and links, then nesting those under VisOptions
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #

vis_options = (
    pd.merge(
        measure_mapping,
        measure_trend,
        how = "left"
    )
    .merge(
        measure_links,
        how = "left"
    )
    .groupby(
        [
            "IndicatorID",
            "MeasureID"
        ],
        dropna = False
    )
    .apply(lambda x: x[[
        "Map", 
        "Trend", 
        "Links"
    ]].to_dict('records'))
    .reset_index()
    .rename(columns = {0: "VisOptions"})
)


#-----------------------------------------------------------------------------------------#
# nesting geotypes
#-----------------------------------------------------------------------------------------#

measure_geotypes = (
    EXP_metadata
    .loc[:, 
        [
            "IndicatorID",
            "IndicatorName",
            "IndicatorLabel",
            "IndicatorDescription",
            "MeasureID",
            "MeasureName",
            "MeasurementType",
            "how_calculated",
            "Sources",
            "DisplayType",
            "GeoType",
            "GeoTypeDescription"
        ]
    ]
    .drop_duplicates()
    .groupby(
        [
            "IndicatorID",
            "IndicatorName",
            "IndicatorLabel",
            "IndicatorDescription",
            "MeasureID",
            "MeasureName",
            "MeasurementType",
            "how_calculated",
            "Sources",
            "DisplayType"
        ],
        dropna = False
    )
    .apply(lambda x: x[["GeoType", "GeoTypeDescription"]].to_dict("records"))
    .reset_index()
    .rename(columns = {0: "AvailableGeoTypes"})
)


#-----------------------------------------------------------------------------------------#
# nesting times
#-----------------------------------------------------------------------------------------#

measure_times = (
    EXP_metadata
    .loc[:, 
        [
            "IndicatorID",
            "IndicatorName",
            "IndicatorLabel",
            "IndicatorDescription",
            "MeasureID",
            "MeasureName",
            "MeasurementType",
            "how_calculated",
            "Sources",
            "DisplayType",
            "TimeDescription",
            "start_period",
            "end_period"
        ]
    ]
    .drop_duplicates()
    .groupby(
        [
            "IndicatorID",
            "IndicatorName",
            "IndicatorLabel",
            "IndicatorDescription",
            "MeasureID",
            "MeasureName",
            "MeasurementType",
            "how_calculated",
            "Sources",
            "DisplayType"
        ],
        dropna = False
    )
    .apply(lambda x: x[["TimeDescription", "start_period", "end_period"]].to_dict("records"))
    .reset_index()
    .rename(columns = {0: "AvailableTimes"})
)


#-----------------------------------------------------------------------------------------#
# combining geotype, times, and vis options, then nesting those under other measure-level info vars
#-----------------------------------------------------------------------------------------#

metadata = (
    pd.merge(
        measure_geotypes,
        measure_times,
        how = "left"
    )
    .merge(
        vis_options,
        how = "left"
    )
    .groupby(
        [
            "IndicatorID",
            "IndicatorName",
            "IndicatorLabel",
            "IndicatorDescription"
        ],
        dropna = False
    )
    .apply(lambda x: x[[
        "MeasureID", 
        "MeasureName", 
        "MeasurementType", 
        "how_calculated",
        "Sources",
        "DisplayType",
        "AvailableGeoTypes",
        "AvailableTimes",
        "VisOptions"
    ]].to_dict('records'))
    .reset_index()
    .rename(columns = {0: "Measures"})
    .merge(
        indicator_comparisons,
        how = "left"
    )
    .loc[:, 
        [
            "IndicatorID",
            "IndicatorName",
            "IndicatorLabel",
            "IndicatorDescription",
            "Comparisons",
            "Measures"
        ]
    ]
    
)


#-----------------------------------------------------------------------------------------#
# saving files
#-----------------------------------------------------------------------------------------#

metadata.to_json(base_dir + "/indicators/metadata/metadata_pretty.json", orient = "records", indent = 2)
metadata.to_json(base_dir + "/indicators/metadata/metadata.json", orient = "records", indent = 0)

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# #
# #                             ---- THIS IS THE END! ----
# #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
